helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java

```diff
@@ -146,8 +146,13 @@ public boolean finishJob(String job) {
     }
     // Add finished job's successors to ready-list
     if (_isJobQueue) {
-      if (_lastJob != null && _parentsToChildren.containsKey(_lastJob)) {
-        _readyJobList.offer(_parentsToChildren.get(_lastJob).iterator().next());
+      while (_lastJob != null && _parentsToChildren.containsKey(_lastJob)) {
+        String nextJob = _parentsToChildren.get(_lastJob).iterator().next();
+        if (!_readyJobList.contains(nextJob)) {
```

Comment:
What's the time complexity of this operation?

Reply:
I understand you concern about time complexity. However, if you look at the whole structure of the code, finish job is only called once and only if the job is finished. Also parallel jobs is not usually set to a large value, I think we are safe here. Also note that we need that for correctness of TF.

Reply:
This is within a while loop. How many times this check will be run?

Reply:
Usually in normal cases where we do not have controller switch we don't run this loop multiple times because _lastJob is keeping us at the end for the list. Also if we have controller switch, it will help us in the long run because we do not need to loop through the jobs that has been scheduled before multiple times.